pkg/nsx/services/securitypolicy/cleanup.go

```diff
@@ -0,0 +1,174 @@
+	return service.ruleBuilder.PagingDeleteResources(ctx, rules, common.DefaultHAPIChildrenCount, service.NSXClient, func(deletedObjs []*model.Rule) {
+		service.ruleStore.DeleteMultipleObjects(deletedObjs)
+	})
+}
+
+// cleanupRulesByVPC is deleting all the NSX groups in the given vpcPath on NSX and/or in local cache.
+// If vpcPath is not empty, the function is called with auto-created VPC case, so it only deletes in the local cache for
+// the NSX resources are already removed when VPC is deleted recursively. Otherwise, it should delete all cached groups
+// on NSX and in local cache.
+func (service *SecurityPolicyService) cleanupGroupsByVPC(ctx context.Context, vpcPath string) error {
+	if vpcPath != "" {
+		groups := service.groupStore.GetByIndex(common.IndexByVPCPathFuncKey, vpcPath)
+		if len(groups) == 0 {
+			return nil
+		}
+		// Delete resources from the store and return.
+		service.groupStore.DeleteMultipleObjects(groups)
+		return nil
+	}
+
+	return cleanGroups(ctx, service.groupStore, service.groupBuilder, service.NSXClient)
+}
+
+// CleanupInfraResources is to clean up the resources created by SecurityPolicyService under path /infra.
+func (service *SecurityPolicyService) CleanupInfraResources(ctx context.Context) error {
+	for _, config := range []struct {
+		store   *ShareStore
+		builder *common.PolicyTreeBuilder[*model.Share]
+	}{
+		{
+			store:   service.projectShareStore,
+			builder: service.projectShareBuilder,
+		}, {
+			store:   service.infraShareStore,
+			builder: service.infraShareBuilder,
+		},
+	} {
+		if err := cleanShares(ctx, config.store, config.builder, service.NSXClient); err != nil {
+			return err
+		}
+	}
+	for _, config := range []struct {
+		store   *GroupStore
+		builder *common.PolicyTreeBuilder[*model.Group]
+	}{
+		{
+			store:   service.projectGroupStore,
+			builder: service.projectGroupBuilder,
+		}, {
+			store:   service.infraGroupStore,
+			builder: service.infraGroupBuilder,
+		},
+	} {
+		if err := cleanGroups(ctx, config.store, config.builder, service.NSXClient); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanShares(ctx context.Context, store *ShareStore, builder *common.PolicyTreeBuilder[*model.Share], nsxClient *nsx.Client) error {
+	cachedObjs := store.List()
+	if len(cachedObjs) == 0 {
+		return nil
+	}
+	log.Info("Cleaning up Shares", "Count", len(cachedObjs))
+	cachedShares := make([]*model.Share, 0)
+	for _, obj := range cachedObjs {
+		share := obj.(*model.Share)
+		share.MarkedForDelete = &MarkedForDelete
+		cachedShares = append(cachedShares, share)
+	}
+
+	return builder.PagingDeleteResources(ctx, cachedShares, common.DefaultHAPIChildrenCount, nsxClient, func(deletedObjs []*model.Share) {
+		store.DeleteMultipleObjects(deletedObjs)
+	})
+}
+
+func cleanGroups(ctx context.Context, store *GroupStore, builder *common.PolicyTreeBuilder[*model.Group], nsxClient *nsx.Client) error {
+	cachedObjs := store.List()
+	log.Info("Cleaning up Groups", "Count", len(cachedObjs))
```

Comment:
Also check cachedObjs length before moving on like this?
https://github.com/vmware-tanzu/nsx-operator/blob/0f6c0d5779200a4e5eb9fdf8dc932d48bf3f3371/pkg/nsx/services/securitypolicy/cleanup.go#L145-L147